src/components/settings/PushNotifications/GlobalPushNotifications.tsx

```diff
@@ -0,0 +1,346 @@
+        return {
+          ...acc,
+          [chainId]: safeAddresses,
+        }
+      }, {})
+    })
+  }
+
+  const totalSignaturesRequired = selectedChains.filter((chainId) => {
+    return !currentNotifiedChains.includes(chainId)
+  }).length
+
+  // Whether Safes need to be (un-)registered with the service
+  const shouldRegisterSelectedSafes = Object.entries(selectedSafes).some(([chainId, safeAddresses]) => {
+    return safeAddresses.some((safeAddress) => !currentNotifiedSafes?.[chainId]?.includes(safeAddress))
+  })
+
+  const shouldUnregisterUnselectedSafes =
+    currentNotifiedSafes &&
+    Object.entries(currentNotifiedSafes).some(([chainId, safeAddresses]) => {
+      return safeAddresses.some((safeAddress) => !selectedSafes[chainId]?.includes(safeAddress))
+    })
+
+  const canSave = shouldRegisterSelectedSafes || shouldUnregisterUnselectedSafes
+
+  const onSave = async () => {
+    if (!canSave) {
+      return
+    }
+
+    const isGranted = await requestNotificationPermission()
+
+    if (!isGranted) {
+      return
+    }
+
+    const registrationPromises: Array<Promise<void>> = []
+
+    const safesToRegister = Object.entries(selectedSafes).reduce<NotifiableSafes>((acc, [chainId, safeAddresses]) => {
+      const safesToRegisterOnChain = safeAddresses.filter(
+        (safeAddress) => !currentNotifiedSafes?.[chainId]?.includes(safeAddress),
+      )
+
+      if (safesToRegisterOnChain.length > 0) {
+        acc[chainId] = safeAddresses
+      }
+
+      return acc
+    }, {})
+
+    const shouldRegister = Object.values(safesToRegister).some((safeAddresses) => safeAddresses.length > 0)
+
+    if (shouldRegister) {
+      registrationPromises.push(registerNotifications(safesToRegister))
+    }
+
+    const shouldUnregister =
+      currentNotifiedSafes &&
+      Object.entries(currentNotifiedSafes).reduce<NotifiableSafes>((acc, [chainId, safeAddresses]) => {
+        const safesToUnregisterOnChain = safeAddresses.filter(
+          (safeAddress) => !selectedSafes[chainId]?.includes(safeAddress),
+        )
+
+        if (safesToUnregisterOnChain.length > 0) {
+          acc[chainId] = safeAddresses
+        }
+        return acc
+      }, {})
+
+    if (shouldUnregister) {
+      for (const [chainId, safeAddresses] of Object.entries(shouldUnregister)) {
+        const shouldUnregsiterDevice =
+          safeAddresses.length === currentNotifiedSafes[chainId].length &&
+          safeAddresses.every((safeAddress) => {
+            return currentNotifiedSafes[chainId]?.includes(safeAddress)
+          })
+
+        if (shouldUnregsiterDevice) {
+          registrationPromises.push(unregisterChainNotifications(chainId))
+          continue
+        }
+
+        for (const safeAddress of safeAddresses) {
+          registrationPromises.push(unregisterSafeNotifications(chainId, safeAddress))
+        }
+      }
+    }
+
+    await Promise.all(registrationPromises)
+
+    trackEvent(PUSH_NOTIFICATION_EVENTS.SAVE_SETTINGS)
+  }
```

Comment:
A lot is going on here and it's a bit hard to follow. Especially with regards to what functions need the state of the component and which are pure. I would suggest to extract as much logic as you can into pure functions outside of the component and keep JSX separate from logic.

Reply:
I've abstracted the code outside of the component in 75b5a63d8189dbe37b42ffb469b1bb2361a2d28a.